datafusion/core/src/physical_plan/file_format/delimited_stream.rs

```diff
@@ -138,29 +138,37 @@ where
 {
     let delimiter = LineDelimiter::new();
 
-    futures::stream::unfold((s, delimiter), |(mut s, mut delimiter)| async move {
-        loop {
-            if let Some(next) = delimiter.next() {
-                return Some((Ok(next), (s, delimiter)));
-            }
+    futures::stream::unfold(
+        (s, delimiter, false),
+        |(mut s, mut delimiter, mut exhausted)| async move {
+            loop {
+                if let Some(next) = delimiter.next() {
+                    return Some((Ok(next), (s, delimiter, exhausted)));
+                } else if exhausted {
+                    return None;
+                }
 
-            match s.next().await {
-                Some(Ok(bytes)) => delimiter.push(bytes),
-                Some(Err(e)) => return Some((Err(e), (s, delimiter))),
-                None => match delimiter.finish() {
-                    Ok(true) => return None,
-                    Ok(false) => continue,
-                    Err(e) => return Some((Err(e), (s, delimiter))),
-                },
+                match s.next().await {
+                    Some(Ok(bytes)) => delimiter.push(bytes),
+                    Some(Err(e)) => return Some((Err(e), (s, delimiter, exhausted))),
+                    None => {
+                        exhausted = true;
```

Comment:
I see this is to handle the case of a missing null terminator, for which there is a test added below :+1: